### PR TITLE
If ioctl

### PR DESCRIPTION
**Pull request by bsdjhb**
This is a different take on ABI compatibility for network related ioctls.  Rather than pushing down ABI-awareness into device drivers and pseudo interfaces, this follows the approach generally used in various device drivers for ABI compat by using a thunk structure to convert the non-native ABI structures (in this case ifreq32 and ifreq64) to a native structure in ifioctl() and always passing native requests down to lower layers.  In particular, this permits removing all of our local diffs from network device drivers and many of the local diffs in pseudo network interface drivers (we still have local diffs in some due to iovec macro changes or due to the cloners taking a __capability pointer for the cloning params).

Some of the ioctls only used by a single pseudo interface should perhaps be pushed back into the pseudo interfaces, but it's also true that we may not really want/need freebsd64 compat for those anyway.

I also think that the current ifr_data_*() and ifr_buffer_*() routines are inherently flawed.  The kernel itself is permitted to issue native ioctls as part of implementing a compat syscall (for example, for CloudABI or Linux compat, etc.) and the ABI of the current thread doesn't matter.  If we wanted to keep ifr_data_get_pointer() for example, I would only keep it for drivers that want to provide compat for more obscure ioctls (e.g. if_ipsec.c or the like), and would change it to add the 'cmd' as an additional argument.  It would then switch on the parameter size instead of SV_PROC_ABI() to determine what type of ifr pointer was passed in.

I've left bpf.c alone for now, and it is the last remaining user of CASE_IOC_IFREQ() because we have other changes in bpf.c that I think we would revert if we choose to go this direction for compat instead of the current approach.

I've also chosen for now to use some tighter bounds on ifr_addr for some of the ioctls, but we could also just dispense with that and always use unbounded caps which is what we do today which would allow a few of the cases to be collapsed.

### CODE REVIEW

File: sys/dev/bxe/bxe.c [hunk 1]
```diff
@@ -4519,29 +4519,34 @@ bxe_ioctl(if_t ifp,
...
         atomic_store_rel_int((volatile unsigned int *)&sc->mtu,
-                             ifr_mtu_get(ifr));
-	if_setmtu(ifp, ifr_mtu_get(ifr));
+                             (unsigned long)ifr->ifr_mtu);
+	/* 
+        atomic_store_rel_long((volatile unsigned long *)&if_getmtu(ifp),
+                              (unsigned long)ifr->ifr_mtu);
+	XXX - Not sure why it needs to be atomic
+	*/
+	if_setmtu(ifp, ifr->ifr_mtu);
```
Comment (jrtc27) [inline sys/dev/bxe/bxe.c:4545]: Hmmmmmmmm

Comment (bsdjhb) [reply sys/dev/bxe/bxe.c:4545]: That is just restoring the upstream code, it's definitely wonky upstream.

File: sys/net/if.c [hunk 2]
```diff
@@ -3549,119 +3032,442 @@ ifioctl(struct socket *so, u_long cmd, caddr_t data, struct thread *td)
...
+		case _IOC_NEWTYPE(IPSECGREQID, struct ifreq32):
+		case _IOC_NEWTYPE(IPSECSREQID, struct ifreq32):
+			thunk.ifr.ifr_data = __USER_CAP(ifr32->ifr_data,
+			    sizeof(uint32_t));
+			break;
+		}
+		saved_cmd = cmd;
+		saved_data = data;
+		data = (caddr_t)&thunk.ifr;
+		cmd = _IOC_NEWTYPE(cmd, struct ifreq);
```
Comment (brooksdavis) [inline sys/net/if.c:3226]: I keep thinking the translation should happen before the inner case statement, but can see it going either way.

**Review by brooksdavis — COMMENTED** [review]
I generally like this. I didn't go this route initially because I wasn't sure that ifioctl was a sufficiently complete funnel.

**Comment by brooksdavis** [comment]
> I also think that the current ifr_data_() and ifr_buffer_() routines are inherently flawed. The kernel itself is permitted to issue native ioctls as part of implementing a compat syscall (for example, for CloudABI or Linux compat, etc.) and the ABI of the current thread doesn't matter. If we wanted to keep ifr_data_get_pointer() for example, I would only keep it for drivers that want to provide compat for more obscure ioctls (e.g. if_ipsec.c or the like), and would change it to add the 'cmd' as an additional argument. It would then switch on the parameter size instead of SV_PROC_ABI() to determine what type of ifr pointer was passed in.

Switching to pass the cmd makes sense. I'm not sure about the in-kernal issue in practice. Are there existing compat cases where the data doesn't require translation (and thus the implementation needs magic to figure out that it's happened).

File: sys/net/if.c (hunk 2 above)
Comment (bsdjhb) [reply sys/net/if.c:3226]: Yeah, I considered that.  I'm not sure which is cleaner.  I could do either way.

**Comment by bsdjhb** [comment]
> I generally like this. I didn't go this route initially because I wasn't sure that ifioctl was a sufficiently complete funnel.

We also couldn't do this originally because freebsd64 was still the native ABI and we would have had to have touch all the drivers anyway to deal with an ifreq_native, etc. which would have been just as invasive.

**Comment by bsdjhb** [comment]
> > I also think that the current ifr_data_() and ifr_buffer_() routines are inherently flawed. The kernel itself is permitted to issue native ioctls as part of implementing a compat syscall (for example, for CloudABI or Linux compat, etc.) and the ABI of the current thread doesn't matter. If we wanted to keep ifr_data_get_pointer() for example, I would only keep it for drivers that want to provide compat for more obscure ioctls (e.g. if_ipsec.c or the like), and would change it to add the 'cmd' as an additional argument. It would then switch on the parameter size instead of SV_PROC_ABI() to determine what type of ifr pointer was passed in.
> 
> Switching to pass the cmd makes sense. I'm not sure about the in-kernal issue in practice. Are there existing compat cases where the data doesn't require translation (and thus the implementation needs magic to figure out that it's happened).

Well, what happened before is that we always followed the process ABI and so we both stored and read the pointer from the "wrong" place in the native ifreq.  That was also less broken than it sounds in practice since the offset of the union doesn't change, so the offset of ifr_addr_get() was always the same for pure cap and 64-bit anyway.  The in-kernel cases right now are mostly things like `SIOCSIFFLAGS`, but the Linux compat really needs to be reworked to use kern_ioctl in which case we would grow several more in-kernel native ioctls.

There are some cases where the data being copied in/out uses fixed size types, etc. such as the GI2C several drivers implement as well as the carp and pfsync ioctls which use a fixed size structure.  Some others though would need to know about the size, and there I think the best option is to pass the raw `cmd` to the driver so it can use the raw `cmd` to decide not only the size/type of ifr_addr, but also the size and layout of the structure to copy in and/or out.

I will update ifr_addr_get_pointer() to take 'cmd' and then fix move some of the compat shims back into a few of the drivers where they are single purpose and best isolated to that driver.

**Comment by bsdjhb** [comment]
> I will update ifr_addr_get_pointer() to take 'cmd' and then fix move some of the compat shims back into a few of the drivers where they are single purpose and best isolated to that driver.

I've done this now.  I still use ifr_addr directly in places that don't support any compatibility shims.  For ifr_buffer all of the compat shims are in ifioctl() so the handlers that use it use direct access.

**Comment by bsdjhb** [comment]
This now uses IFREQ64, it also has kept accessors for ifr_addr and ifr_buffer but adds a 'cmd' argument to them.  All of the shims for ifreq32 I had added previously are gone since I hadn't realized that sizeof(ifreq32) == sizeof(ifreq64).  Compared to the original change, this pushes a few more one-off driver-private compat handling back into some pseudo-interfaces.  It also keeps some compat handling for driver-private ioctls I had previously removed.  As a result, a handful more files now contain local changes still.

**Pull request merged**